data/ledger.go

```diff
@@ -308,15 +308,53 @@ func (l *Ledger) ConsensusParams(r basics.Round) (config.ConsensusParams, error)
 }
 
 // ConsensusVersion gives the consensus version agreed on in a given round,
-// returning an error if we don't have that round or we have an
-// I/O error.
+// returning an error if the consensus version could not be figured using
+// either the block header for the given round, or the latest block header.
 // Implements agreement.Ledger.ConsensusVersion
 func (l *Ledger) ConsensusVersion(r basics.Round) (protocol.ConsensusVersion, error) {
 	blockhdr, err := l.BlockHdr(r)
-	if err != nil {
+	if err == nil {
+		return blockhdr.UpgradeState.CurrentProtocol, nil
+	}
+	// try to see if we can figure out what the version would be.
+	latestRound := l.Latest()
+	// if the request round was for an older round, then just say the we don't know.
+	if r < latestRound {
 		return "", err
 	}
-	return blockhdr.UpgradeState.CurrentProtocol, nil
+	// the request was for a future round. See if we have any known plans for the next round.
+	latestBlockhdr, err := l.BlockHdr(latestRound)
+	// if we have the lastest block header, look inside and try to figure out if we can deduce the
+	// protocol version for the given round.
+	if err == nil {
+		// check to see if we have a protocol upgrade.
+		if latestBlockhdr.NextProtocolSwitchOn == 0 {
+			// no protocol upgrade taking place, we have *at least* UpgradeVoteRounds before the protocol version would get changed.
+			// it's safe to ignore the error case here since we know that we couldn't reached to this "known" round
+			// without having the binary supporting this protocol version.
+			currentConsensusParams, _ := config.Consensus[latestBlockhdr.CurrentProtocol]
+			// we're using <= here since there is no current upgrade on this round, and if there will be one on the subsequent round
+			// it would still be correct until (latestBlockhdr.Round + currentConsensusParams.UpgradeVoteRounds)
+			if r <= latestBlockhdr.Round+basics.Round(currentConsensusParams.UpgradeVoteRounds) {
```

Comment:
Should MinUpgradeWaitRounds also be taken into account here? i.e.
currentConsensusParams.UpgradeVoteRounds + MinUpgradeWaitRounds?
It is probably not important, I can't think of why one would check too much into the future.

Reply:
I don't think that it really matter, and if we go this path it's becoming really complicated.
You see, in order to figure that out, we need to know which protocol we're going to upgrade *to* ( and there could be several ! ). Then, depending on the upgrade path, we find what the delay is. If the delay is zero, we're going to default to `DefaultUpgradeWaitRounds`, otherwise, the upgrade is bounded to [min..max].
I think that if we will need it in the future, we will always be able to implement it. Right now, we just need the `current round + 1`